Add nonpareil

jfy133 opened the pull request:
Add nonpareil metagenome redundancy estimation

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/taxprofiler/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/taxprofiler _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

jfy133 commented:
TODO: add nonpareil R based visualisation support

jfy133 commented on subworkflows/local/nonpareil.nf:
```diff
@@ -0,0 +1,58 @@
+    NONPAREIL_SET ( NONPAREIL_NONPAREIL.out.npo.map {meta, npo -> [[id: 'all'], npo] }.groupTuple() )
+
+    ch_versions = ch_versions.mix( NONPAREIL_NONPAREIL.out.versions.first(), NONPAREIL_CURVE.out.versions.first(), NONPAREIL_SET.out.versions.first() )
+
```
```suggestion
```

jfy133 commented on nextflow_schema.json:
```diff
@@ -258,6 +258,29 @@
+            "properties": {
+                "perform_shortread_redundancyestimation": {
+                    "type": "boolean",
+                    "description": "Turn on metagenome sequencing redundancy estimation with nonpareil",
```
```suggestion
                    "description": "Turn on short-read metagenome sequencing redundancy estimation with nonpareil. Warning: only use for shallow short-read sequencing datasets.",
```

jfy133 commented on docs/usage.md:
```diff
@@ -247,6 +248,39 @@ There is currently one option for long-read Oxford Nanopore processing: [`porech
+Nonpareil is performed on processed reads (i.e. after fastp or AdapterRemoval). This will run on either the first read of each read pair (as recommended by the authors), or on merged reads.
+
+Before using this tool please note the following caveats:
+
+- It is not recommended to run this on deep sequencing data, or very large datasets
+  - Nonpareil requires uncompressed FASTQ files, and nf-core/taxprofiler will uncompress these in your working directory, potentially with a extremely large hard-drive footprint.
+- Your shortest reads _after_ processing should not go below 24bp (see warning below)
+- It is not recommended to keep unmerged (`--shortread_qc_includeunmerged`) reads when using the calculation.
+
```
```suggestion

:::warning
- It is not recommended to run this on deep sequencing data, or very large datasets
  - Nonpareil requires uncompressed FASTQ files, and nf-core/taxprofiler will uncompress these in your working directory, potentially with a extremely large hard-drive footprint.
- Your shortest reads _after_ processing should not go below 24bp (see warning below)
- It is not recommended to keep unmerged (`--shortread_qc_includeunmerged`) reads when using the calculation.

```
:::

jfy133 commented on docs/usage.md:
```diff
@@ -247,6 +248,39 @@ There is currently one option for long-read Oxford Nanopore processing: [`porech
+- Your shortest reads _after_ processing should not go below 24bp (see warning below)
+- It is not recommended to keep unmerged (`--shortread_qc_includeunmerged`) reads when using the calculation.
+
+:::warning
```
```suggestion
:::info
```

jfy133 commented:
@nf-core-bot fix linting

LilyAnderssonLee commented on docs/usage.md:
```diff
@@ -247,6 +248,41 @@ There is currently one option for long-read Oxford Nanopore processing: [`porech
+
+Metagenome 'coverage' or sequencing complexity estimations of short-read datasets can be activated with `--perform_shortread_redundancyestimation`.
+
+This turns on checking of read redundancy in a sequencing library using [Nonpareil](https://nonpareil.readthedocs.io/en/latest/), to provide an estimation of whether you have sequenced enough to capture all possible genomes present in your metagenomic (with the assumption that once you've sequenced enough, you will keep sequencing PCR amplicons rather than unique reads).
```
```suggestion
This turns on checking of read redundancy in a sequencing library using [Nonpareil](https://nonpareil.readthedocs.io/en/latest/), to provide an estimation of whether you have sequenced enough to capture all possible genomes present in your metagenomic sample (with the assumption that once you've sequenced enough, you will keep sequencing PCR amplicons rather than unique reads).
```

LilyAnderssonLee commented on subworkflows/local/nonpareil.nf:
```diff
@@ -0,0 +1,57 @@
+*/
+def extractNonpareilExtensionFromArrays(ch_input) {
+
+return ch_profile
+.map { meta, profile -> [meta.db_name, meta, profile] }
```
Please format the function.

jfy133 commented on conf/test_falco.config:
I'm not sure I follow?

LilyAnderssonLee replied:
I meant that should we also include `perform_shortread_redundancyestimation = false` in this config file.

jfy133 commented on subworkflows/local/nonpareil.nf:
```diff
@@ -0,0 +1,57 @@
+def extractNonpareilExtensionFromArrays(ch_input) {
+
+return ch_profile
+.map { meta, profile -> [meta.db_name, meta, profile] }
+    .combine(ch_database, by: 0)
```
```suggestion
    .map { meta, profile -> [meta.db_name, meta, profile] }
    .combine(ch_database, by: 0)
```

LilyAnderssonLee reviewed (approved):
I ran the test locally, and the results look good.

Midnighter commented:
Is it the linting that touched so many files? It's a bit confusing, but Lili already approved so that's also fine 🙂 .

jfy133 commented:
> Is it the linting that touched so many files? It's a bit confusing, but Lili already approved so that's also fine 🙂 .

This is nf-test for you... I added 3 modules and updated one.

Each comes with the typical ~3 module files, but for each module an additional 3-4 for nf-test and the conda env file, so doubling each. And then I updated quite a few test configs 🤷‍♀️![Screenshot_20231130-150252.png](https://github.com/nf-core/taxprofiler/assets/17950287/b5c9de07-74ff-43bf-bbbd-edffd29caaac)

jfy133 commented:
Thanks @LilyAnderssonLee !
